process_report/process_report.py

```diff
@@ -1,9 +1,106 @@
 import argparse
 import os
+import sys
 
 import pandas
 
 
+### Invoice field names
+PROJECT_FIELD = 'Project - Allocation'
+MANAGER_PI_FIELD = 'Manager (PI)'
+INSTITUTION_FIELD = 'Institution'
+COST_FIELD = 'Cost'
+CREDIT_FIELD = 'Credit'
+CREDIT_CODE_FIELD = 'Credit Code'
+BALANCE_FIELD = 'Balance'
+###
+
+
+def apply_credits_0001(dataframe):
+    credit_code = "0001"
+    pass
+
+
+def apply_credits_0002(dataframe):
+    """Applies the New PI Credit. This credit function expects the 
+    env var `C0002_OLD_PI` to be set, pointing to a txt file containing old PIs"""
+    credit_code = "0002"
+    credit_amount = 1000
+
+    old_pi_list = set()
+    try:
+        with open(os.getenv("C0002_OLD_PI")) as f:
+            for pi in f: old_pi_list.add(pi.strip())
+
+    except Exception:
+        print("Old PI file does not exist, or the C0002_OLD_PI env var is not set")
+        sys.exit(1)
+
+    print("Old pi list: ", old_pi_list)
+    pi_list = dataframe[MANAGER_PI_FIELD].unique()
+
+    for pi in pi_list:
+        if pi != pi: continue   # NaN check
+        if pi in old_pi_list: continue  # Is the PI an old PI?
+
+        pi_projects = dataframe[dataframe[MANAGER_PI_FIELD] == pi]
+        rem_credit = credit_amount
+        for i, row in pi_projects.iterrows():
+            project_cost = row[COST_FIELD]
+            if project_cost >= rem_credit:
+                dataframe.at[i, CREDIT_FIELD] =  rem_credit
+                dataframe.at[i, CREDIT_CODE_FIELD] = credit_code
+                dataframe.at[i, BALANCE_FIELD] = row[COST_FIELD] - rem_credit
+                break
+            else:
+                dataframe.at[i, CREDIT_FIELD] =  project_cost
+                dataframe.at[i, CREDIT_CODE_FIELD] = credit_code
+                dataframe.at[i, BALANCE_FIELD] = 0
+                rem_credit -= project_cost
+    
+    return dataframe
+
+
+applied_credits = [apply_credits_0002]
+
+
+def get_institution_from_pi(pi_uname):
+    
+    institute_map = {
+        "northeastern.edu"      : "Northeastern University",
+        "bu.edu"                : "Boston University",
+        "bentley.edu"           : "Bentley",
+        "uri.edu"               : "University of Rhode Island",
+        "redhat.com"            : "Red Hat",
+        "childrens.harvard.edu" : "Boston Childrens Hospital",
+        "mclean.harvard.edu"    : "McLean Hospital",
+        "meei.harvard.edu"      : "Massachusetts Eye & Ear",
+        "dfci.harvard.edu"      : "Dana-Farber Cancer Institute",
+        "bwh.harvard.edu"       : "Brigham and Women's Hospital",
+        "bidmc.harvard.edu"     : "Beth Israel Deaconess Medical Center",
+        "harvard.edu"           : "Harvard University",
```

Comment:
I do not know how Python prioritizes matches, since any of the 6 items above this line could match to this row how do we make sure that any of the 6 above match their value, and then only harverad.edu matches not in the list above becomes Harvard University?

Reply:
@joachimweyl From what I found, Python dictionary is [ordered](https://docs.python.org/3/library/stdtypes.html#dict.values), so the row will be matched against the Harvard affiliates before being matched against `harvard.edu`. I added this scenario in my [test case ](https://github.com/CCI-MOC/process_csv_report/pull/12/files#diff-2ab8a00ce6228ed90b8929a4e45622b1586da07acbc85aa2d71533747f594ca9R188) to make sure.